### ISSUE TITLE
False positive for "no overloads match the provided argument" inside a function

### ISSUE THREAD

**Issue by ipkiss42** [microsoft/pyright#1565]
**Describe the bug**
I get the `no overloads for "fspath" match the provided argument` when the code is inside a function but not outside.

**To Reproduce**
Sample code:

```python
import os
from typing import Union

MyPathLike = Union[str, os.PathLike]

def func(path: MyPathLike) -> None:
    os.fspath(path)

foo: MyPathLike = "foo"

os.fspath(foo)
func(foo)
```

**Expected behavior**
No error

**Screenshots or Code**
Output:
```
$ pyright
No configuration file found.
stubPath /tmp/pyright5/root/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/pyright5/root/src/repro.py
  7:5 - error: No overloads for "fspath" match the provided arguments
    Argument types: (MyPathLike) (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
Completed in 0.671sec
```

Interestingly, Pyright complains about the usage at line 7 (inside the function) but not at line 11, even though they should be equivalent.

**VS Code extension or command-line**
```
$ pyright --version                                                                                          1
pyright 1.1.117
```

**Comment by vemel** [comment]
Actually, the issue is different. It works for line 11, because `foo` has type `Literal["foo"]`, and `os.fspath` accepts `str`

For line 7, foo is `Union[str, os.PathLike]`. While `os.fspath` has overloads for both `str` and `PathLike`, it does not have an overload for `Union[str, PathLike]`, that is why you get an error.

Common sense says that if overloads exist for all types in `Union`, there should be no type error. Sounds like a feature request.

**Comment by erictraut** [comment]
>Common sense says that if overloads exist for all types in Union, there should be no type error.

The current behavior is intended. Pyright does not attempt to match multiple overloads for a single call. This would lead to combinatoric explosion if multiple arguments were union types. While most unions contain just a couple of types, some contain thousands.

Changing this behavior would also be a significant change to pyright's internal type analyzer. I don't have any current plans to change it.

**Comment by vemel** [comment]
Makes sense. If you split types per overload, it works with no issues:

```python
MyPathLike = Union[str, os.PathLike]

def func(path: MyPathLike) -> None:
    if isinstance(path, str):
        os.fspath(path)
    else:
        os.fspath(path)
```

**Comment by ipkiss42** [comment]
> This would lead to combinatoric explosion if multiple arguments were union types. While most unions contain just a couple of types, some contain thousands.

It's not clear to me why each argument couldn't be checked independently, but I'll take your word for it :)

What would be your recommendation for Pyright users running into this issue? Splitting types per overload doesn't sound very doable, in particular if we are talking about thousands of types in the union.

**Comment by erictraut** [comment]
You can't check arguments independently because the types of all arguments need to be considered when selecting an overload. Let's imagine there were 10 overloads (which isn't uncommon in some libraries) with three arguments. If each of those arguments had 1000 union elements, that would require 10^10 type evaluations.

This is how overloads work in other languages (e.g. TypeScript). Only one overload is selected and evaluated. This has never been an issue for me in TypeScript, so I suspect it will come up very rarely in Python as well. As such, I think it's reasonable in these rare cases for manually splitting types. You don't need to split types thousands of times if you know about the types and overloads involved.

**Comment by DevilXD** [comment]
> The current behavior is intended. Pyright does not attempt to match multiple overloads for a single call. This would lead to combinatoric explosion if multiple arguments were union types. While most unions contain just a couple of types, some contain thousands.
> 
> Changing this behavior would also be a significant change to pyright's internal type analyzer. I don't have any current plans to change it.

> You can't check arguments independently because the types of all arguments need to be considered when selecting an overload.
> This is how overloads work in other languages (e.g. TypeScript). Only one overload is selected and evaluated.

@erictraut This issue can be easily solved, by treating the final function definition typing as an additional "overload", AKA a valid call. I'm pretty sure this shouldn't break anything, doesn't lead to "combinatoric explosion" as you've described, and shouldn't require a lot of changes to implement. Right now, it can be worked around by duplicating the final type signature as an overload, but that feels like unneeded boilerplate code.

Side-note: MyPy is currently having the same issue: https://github.com/python/mypy/issues/10194

**Comment by erictraut** [comment]
@DevilXD, pyright used to include the implementation signature, but it was pointed out that PEP 484 specifically disallows this. See this issue: https://github.com/microsoft/pyright/issues/1094.

**Comment by erictraut** [comment]
After receiving a number of bug reports on this issue, I've decided to change the behavior of overload matching to more closely mirror that of other type checkers. New versions of of pyright will perform union expansion and evaluate multiple overloads. I've addressed the combinatoric explosion problem by limiting expansion to 64 combinations. In practice, this will occur very infrequently because it would require more than 6 arguments with two union subtypes each or more than 2 arguments with 8 union subtypes each.

**Comment by erictraut** [comment]
This change is included in pyright 1.1.135, which I just published. It will also be included in the next release of pylance.
